diff --git a/Cabal/src/Distribution/Simple/PreProcess.hs b/Cabal/src/Distribution/Simple/PreProcess.hs
index d47dd8035..da158a5a3 100644
--- a/src/Distribution/Simple/PreProcess.hs
+++ b/src/Distribution/Simple/PreProcess.hs
@@ -464,7 +464,7 @@ ppHsc2hs bi lbi clbi =
     -- Returns a list of command line arguments that can either be passed
     -- directly, or via a response file.
     genPureArgs :: Version -> ConfiguredProgram -> String -> String -> [String]
-    genPureArgs hsc2hsVersion gccProg inFile outFile =
+    genPureArgs hsc2hsVersion gccProg inFile outFile = ordNub $
           -- Additional gcc options
           [ "--cflag=" ++ opt | opt <- programDefaultArgs  gccProg
                                     ++ programOverrideArgs gccProg ]
